File: commons-cli-graal/src/main/java/org/apache/commons/cli/OptionBuilder.java
/*
  Licensed to the Apache Software Foundation (ASF) under one or more
  contributor license agreements.  See the NOTICE file distributed with
  this work for additional information regarding copyright ownership.
  The ASF licenses this file to You under the Apache License, Version 2.0
  (the "License"); you may not use this file except in compliance with
  the License.  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */

package org.apache.commons.cli;

import org.graalvm.polyglot.PolyglotException;
import org.graalvm.polyglot.Value;
/**
 * OptionBuilder allows the user to create Options using descriptive methods.
 * <p>
 * Details on the Builder pattern can be found at
 * <a href="http://c2.com/cgi-bin/wiki?BuilderPattern">http://c2.com/cgi-bin/wiki?BuilderPattern</a>.
 * <p>
 * This class is NOT thread safe. See <a href="https://issues.apache.org/jira/browse/CLI-209">CLI-209</a>
 *
 * @since 1.0
 * @deprecated since 1.3, use {@link Option#builder(String)} instead
 */
@Deprecated
public final class OptionBuilder {

    /** Option builder instance */
    private static final OptionBuilder INSTANCE = new OptionBuilder();

    /** A reference to the python class */
    private static Value clz = ContextInitializer.getPythonClass("option_builder.py", "OptionBuilder");

    /**
     * Creates an Option using the current settings
     *
     * @return the Option instance
     * @throws IllegalArgumentException if {@code longOpt} has not been set.
     */
    public static Option create() throws IllegalArgumentException {
        try {
            Value obj = clz.invokeMember("create");
            return Option.create(obj);
        } catch (PolyglotException e) {
            throw new IllegalArgumentException(e.getMessage());
        }
    }

    /**
     * Creates an Option using the current settings and with the specified Option {@code char}.
     *
     * @param opt the character representation of the Option
     * @return the Option instance
     * @throws IllegalArgumentException if {@code opt} is not a valid character. See Option.
     */
    public static Option create(final char opt) throws IllegalArgumentException {
        return create(String.valueOf(opt));
    }

    /**
     * Creates an Option using the current settings and with the specified Option {@code char}.
     *
     * @param opt the {@code java.lang.String} representation of the Option
     * @return the Option instance
     * @throws IllegalArgumentException if {@code opt} is not a valid character. See Option.
     */
    public static Option create(final String opt) throws IllegalArgumentException {
        try {
            Value obj = clz.invokeMember("create", opt);
            return Option.create(obj);
        } catch (PolyglotException e) {
            throw new IllegalArgumentException(e.getMessage());
        }
    }

    /**
     * The next Option created will require an argument value.
     *
     * @return the OptionBuilder instance
     */
    public static OptionBuilder hasArg() {
        clz.invokeMember("has_arg");

        return INSTANCE;
    }

    /**
     * The next Option created will require an argument value if {@code hasArg} is true.
     *
     * @param hasArg if true then the Option has an argument value
     * @return the OptionBuilder instance
     */
    public static OptionBuilder hasArg(final boolean hasArg) {
        clz.invokeMember("has_arg", hasArg);

        return INSTANCE;
    }

    /**
     * The next Option created can have unlimited argument values.
     *
     * @return the OptionBuilder instance
     */
    public static OptionBuilder hasArgs() {
        clz.invokeMember("has_args");

        return INSTANCE;
    }

    /**
     * The next Option created can have {@code num} argument values.
     *
     * @param num the number of args that the option can have
     * @return the OptionBuilder instance
     */
    public static OptionBuilder hasArgs(final int num) {
        clz.invokeMember("has_args", num);

        return INSTANCE;
    }

    /**
     * The next Option can have an optional argument.
     *
     * @return the OptionBuilder instance
     */
    public static OptionBuilder hasOptionalArg() {
        clz.invokeMember("has_optional_arg");

        return INSTANCE;
    }

    /**
     * The next Option can have an unlimited number of optional arguments.
     *
     * @return the OptionBuilder instance
     */
    public static OptionBuilder hasOptionalArgs() {
        clz.invokeMember("has_optional_args");

        return INSTANCE;
    }

    /**
     * The next Option can have the specified number of optional arguments.
     *
     * @param numArgs - the maximum number of optional arguments the next Option created can have.
     * @return the OptionBuilder instance
     */
    public static OptionBuilder hasOptionalArgs(final int numArgs) {
        clz.invokeMember("has_optional_args", numArgs);

        return INSTANCE;
    }

    /**
     * The next Option created will be required.
     *
     * @return the OptionBuilder instance
     */
    public static OptionBuilder isRequired() {
        clz.invokeMember("is_required");

        return INSTANCE;
    }

    /**
     * The next Option created will be required if {@code required} is true.
     *
     * @param newRequired if true then the Option is required
     * @return the OptionBuilder instance
     */
    public static OptionBuilder isRequired(final boolean newRequired) {
        clz.invokeMember("is_required", newRequired);

        return INSTANCE;
    }

    /**
     * The next Option created will have the specified argument value name.
     *
     * @param name the name for the argument value
     * @return the OptionBuilder instance
     */
    public static OptionBuilder withArgName(final String name) {
        clz.invokeMember("with_arg_name", name);

        return INSTANCE;
    }

    /**
     * The next Option created will have the specified description
     *
     * @param newDescription a description of the Option's purpose
     * @return the OptionBuilder instance
     */
    public static OptionBuilder withDescription(final String newDescription) {
        clz.invokeMember("with_description", newDescription);

        return INSTANCE;
    }

    /**
     * The next Option created will have the following long option value.
     *
     * @param newLongopt the long option value
     * @return the OptionBuilder instance
     */
    public static OptionBuilder withLongOpt(final String newLongopt) {
        clz.invokeMember("with_long_opt", newLongopt);

        return INSTANCE;
    }

    /**
     * The next Option created will have a value that will be an instance of {@code type}.
     *
     * @param newType the type of the Options argument value
     * @return the OptionBuilder instance
     * @since 1.3
     */
    public static OptionBuilder withType(final Class<?> newType) {
        clz.invokeMember("with_type", newType);

        return INSTANCE;
    }

    /**
     * The next Option created will have a value that will be an instance of {@code type}.
     * <p>
     * <b>Note:</b> this method is kept for binary compatibility and the input type is supposed to be a {@link Class}
     * object.
     *
     * @param newType the type of the Options argument value
     * @return the OptionBuilder instance
     * @deprecated since 1.3, use {@link #withType(Class)} instead
     */
    @Deprecated
    public static OptionBuilder withType(final Object newType) {
        return withType((Class<?>) newType);
    }

    /**
     * The next Option created uses '{@code =}' as a means to separate argument values.
     *
     * <b>Example:</b>
     *
     * <pre>
     * Option opt = OptionBuilder.withValueSeparator().create('D');
     *
     * CommandLine line = parser.parse(args);
     * String propertyName = opt.getValue(0);
     * String propertyValue = opt.getValue(1);
     * </pre>
     *
     * @return the OptionBuilder instance
     */
    public static OptionBuilder withValueSeparator() {
        clz.invokeMember("with_value_separator");

        return INSTANCE;
    }

    /**
     * The next Option created uses {@code sep} as a means to separate argument values.
     * <p>
     * <b>Example:</b>
     *
     * <pre>
     * Option opt = OptionBuilder.withValueSeparator('=').create('D');
     *
     * String args = "-Dkey=value";
     * CommandLine line = parser.parse(args);
     * String propertyName = opt.getValue(0); // will be "key"
     * String propertyValue = opt.getValue(1); // will be "value"
     * </pre>
     *
     * @param sep The value separator to be used for the argument values.
     *
     * @return the OptionBuilder instance
     */
    public static OptionBuilder withValueSeparator(final char sep) {
        clz.invokeMember("with_value_separator", sep);

        return INSTANCE;
    }

    /**
     * private constructor to prevent instances being created
     */
    private OptionBuilder() {
        // hide the constructor
    }
}

File: commons-cli-python/src/main/python/option_builder.py
from main.python.option import Option
from main.python.java_handler import java_handler, Types

@java_handler
class OptionBuilder:
    """
    OptionBuilder allows the user to create Options using descriptive methods that follows builder pattern.
    """

    long_option = None
    description = None
    arg_name = None
    required = False
    arg_count = Option.UNINITIALIZED
    type = Types.String
    optional_arg = False
    value_separator = '\0'
    
    @staticmethod
    def create(opt = None, no_func_arg = True) -> Option:
        """
        Creates an Option using the current settings.
        :param opt: the character representation of the Option
        :return: the Option instance
        :raises: ValueError if long_opt has not been set or opt is not a valid character. See Option.
        """
        if no_func_arg and opt is None:
            if OptionBuilder.long_option is None:
                OptionBuilder.reset()
                raise ValueError("must specify longopt")
            
            return OptionBuilder.create(None, False)
        
        option = None
        try:
            # create the option
            option = Option(opt, OptionBuilder.description)

            # set the option properties
            option.set_long_opt(OptionBuilder.long_option)
            option.set_required(OptionBuilder.required)
            option.set_optional_arg(OptionBuilder.optional_arg)
            option.set_args(OptionBuilder.arg_count)
            option.set_type(OptionBuilder.type)
            option.set_value_separator(OptionBuilder.value_separator)
            option.set_arg_name(OptionBuilder.arg_name)
        finally:
            # reset the OptionBuilder properties
            OptionBuilder.reset()

        # return the Option instance
        return option

    @staticmethod
    def has_arg(has_arg=None):
        """
        The next Option created will require an argument value if 'has_arg' is true.

        If 'has_arg' is None, the next Option will require an argument value.

        :param has_arg: if true then the Option has an argument value
        :return: the OptionBuilder instance
        """
        if has_arg is None:
            OptionBuilder.arg_count = 1
        else:
            OptionBuilder.arg_count = 1 if has_arg else Option.UNINITIALIZED

        return OptionBuilder

    
    @staticmethod
    def has_args(num=None):
        """
        Specifies the number of arguments the next Option created can have.
        If `num` is None, the next Option created can have unlimited argument values.
        
        :param num: Optional; The number of args that the option can have. 
                    If not provided, Option can have unlimited values. 
                    Default is None.
        :return: OptionBuilder instance with arg_count set according to provided `num`.
        """
        if num is None:
            OptionBuilder.arg_count = Option.UNLIMITED_VALUES
        else:
            OptionBuilder.arg_count = num

        return OptionBuilder

    
    @staticmethod
    def has_optional_arg():
        """
        The next Option can have an optional argument.
        
        :return: the OptionBuilder instance
        """
        OptionBuilder.arg_count = 1
        OptionBuilder.optional_arg = True

        return OptionBuilder
    
    @staticmethod
    def has_optional_args(num_args = None):
        """
        Specifies the number of optional arguments the next Option can have. 

        If num_args is not provided, the next Option created can have an unlimited number of optional arguments. 
        If num_args is provided, the next Option created can have up to num_args optional arguments.
        
        :param num_args: Optional; the maximum number of optional arguments the next Option created can have.
        :return: the OptionBuilder instance
        """
        if num_args is None:
            OptionBuilder.arg_count = Option.UNLIMITED_VALUES
            OptionBuilder.optional_arg = True
        else:
            OptionBuilder.arg_count = num_args
            OptionBuilder.optional_arg = True

        return OptionBuilder

        
    @staticmethod
    def is_required(required = None):
        """
        Specifies whether the next Option created will be required or not. 
        
        If required is not provided, the next Option will be required. 
        If required is provided, the next Option will be required if required is true.
        
        :param required: Optional; if true then the Option is required
        :return: the OptionBuilder instance
        """
        if required is None:
            OptionBuilder.required = True
        else:
            OptionBuilder.required = required

        return OptionBuilder


    @staticmethod
    def reset():
        """
        Resets the member variables to their default values.
        """
        OptionBuilder.description = None
        OptionBuilder.arg_name = None
        OptionBuilder.long_option = None
        OptionBuilder.type = Types.String
        OptionBuilder.required = False
        OptionBuilder.arg_count = Option.UNINITIALIZED
        OptionBuilder.optional_arg = False
        OptionBuilder.value_separator = '\0'

    @staticmethod
    def with_arg_name(name: str):
        """
        The next Option created will have the specified argument value name.
        
        :param name: the name for the argument value
        :return: the OptionBuilder instance
        """
        OptionBuilder.arg_name = name

        return OptionBuilder
    
    @staticmethod
    def with_description(new_description: str):
        """
        The next Option created will have the specified description
        
        :param new_description: a description of the Option's purpose
        :return: the OptionBuilder instance
        """
        OptionBuilder.description = new_description

        return OptionBuilder
    
    @staticmethod
    def with_long_opt(new_long_opt: str):
        """
        The next Option created will have the following long option value.
        
        :param new_long_opt: the long option value
        :return: the OptionBuilder instance
        """
        OptionBuilder.long_option = new_long_opt

        return OptionBuilder
    
    @staticmethod
    def with_type(new_type: type):
        """
        The next Option created will have a value that will be an instance of type.
        
        :param new_type: the type of the Options argument value
        :return: the OptionBuilder instance
        """
        OptionBuilder.type = new_type

        return OptionBuilder
    
    # 
    # The translation with Object argument has been omitted
    # as Python does not have the Object type
    #

    @staticmethod
    def with_value_separator(seperator=None):
        """
        Specifies the character to be used as a separator for argument values in the next Option created. 

        If no separator is provided, '=' will be used as the default separator.
        If a separator is provided, that will be used as the separator for argument values.

        :param seperator: Optional; the value separator to be used for the argument values.
        :return: the OptionBuilder instance
        """
        if seperator is None:
            OptionBuilder.value_separator = '='
        else:
            OptionBuilder.value_separator = seperator

        return OptionBuilder

